### ISSUE TITLE
Federation: Support deploying federation control plane on non-cloud environment

### ISSUE THREAD

**Issue by shashidharatd** [kubernetes/kubernetes#40536]
There may be scenarios where there is a need to federate k8s clusters on private-cloud (on-premise) environments.
Listed below are some of the tasks to achieve the scenario above

- [x] Federation control plane deployment itself should be able to be deployed in non-cloud environments. #40516
- [x] Should be possible to disable persistent storage for etcd. #40862
- [x] Ability to configure desired DNS provider. Public DNS servers (AWS Route53/ Google CloudDNS) or Private DNS servers (e.g. CoreDNS) #40528 
- [x] Document Federation CoreDNS support. https://github.com/kubernetes/kubernetes.github.io/pull/2810
- [x] coredns federation provider should support etcd with TLS #47049
- [ ] Document using hosted DNS providers (AWS Route53/ Google CloudDNS) in non-cloud environments.

**Comment by shashidharatd** [comment]
@marun, Please feel free to correct or add any additional points to this issue.

**Comment by shashidharatd** [comment]
This issue will partly address the `deployment` part of federation on non-google clouds as described in #39870

**Comment by marun** [comment]
I think this issue can serve as an umbrella issue to track the issues targeting improved support for non-cloud deployments.  Rather than conflating the work involved, work items can be tracked and discussed separately.  There is already an issue for supporting other ways of exposing the federation api (targeted by #40516): 

 - [[Federation] `kubefed init` should also work on clusters without a service load balancer implementation](https://github.com/kubernetes/kubernetes/issues/39271)

There's already an issue for documenting CoreDNS configuration, which is related to improving support for non-cloud deployment of the federation control plane:

- [Document Federation CoreDNS support.](https://github.com/kubernetes/kubernetes.github.io/issues/2197)

Issues need to be created for making pvc optional (e.g. ``--etcd-pvc``) and supporting dns provider configuration on non-cloud (targeted by #40528).

**Comment by marun** [comment]
cc: @madhusudancs @kubernetes/sig-federation-misc

**Comment by marun** [comment]
I think it should be possible to disable integration with a dns service provider (e.g. ``--dns-provider=none``).  Whether the environment is cloud or non-cloud, I think that having a hard requirement for integration with dnaas imposes a penalty on the development and testing of features that don't depend on dns.

cc: @quinton-hoole @csbell @rrati @madhusudancs @nikhiljindal

**Comment by mattymo** [comment]
https://github.com/kubernetes/kubernetes/issues/42995 should be considered for the list

**Comment by liqlin2015** [comment]
Is there any way we can use a local service load balancer for federation service?

And is there any plan to support services only with external IP? Currently in the `dns.go:wantsDNSRecords()` function, we hard coded to only support `v1.ServiceTypeLoadBalancer` type of services. Can we use external IP as another option for on-prem clusters?

**Comment by shashidharatd** [comment]
what is the meaning of `local service load balancer`.

Currently only loadbalancer type services are supported for federated services. Currently there is no other option available as we know, which works.

**Comment by luckyfengyong** [comment]
@shashidharatd In the doc of federation service discovery, it is mentioned as follows 

<img width="878" alt="image" src="https://user-images.githubusercontent.com/8002217/29038662-81ad5fc8-7b76-11e7-965a-9d1073d86477.png">

After checking the codes we found only when a service is created with the type of Loadbalancer, the DNS record of the service is created in federation dns service. Just wondering what if the service is created with exnternalip. For that case, can we also ask federation dns service to create the DNS record?

Basically our use case is just like what this ticket is described to run cluster federation within on-prem environment. However we find there is no out-of-box external loadbalancer service available.

Therefore we will need either ask dns server of cluster federation supporting the service with externalip, or have a out-of-box loadbalancer service (just like coredns as out-of-box dns server of cluster federation)

Any suggestion?

**Comment by quinton-hoole** [comment]
Required for GA of federation.

**Comment by luxas** [comment]
@quinton-hoole Any relevant design doc/proposal for this issue/feature request?

**Comment by quinton-hoole** [comment]
@luxas Nothing beyond what it linked in this issue, that I am aware of.

**Comment by shashidharatd** [comment]
All the tasks we envisioned are complete and hence closing this issue. If anyone feels this issue needs to be kept open for additional tasks, please feel free to do so.

**Comment by luxas** [comment]
@shashidharatd Where is the documentation for this feature?

**Comment by shashidharatd** [comment]
@luxas, here is the documentation for running federation with CoreDNS as dns provider. https://kubernetes.io/docs/tasks/federation/set-up-coredns-provider-federation.
In non-cloud environments, we could also use hosted DNS providers such as AWS Route53 or Google Cloud-DNS by specifying the dns-provider configuration as secret to `kubefed init`. 

I agree there is no detailed document for user explaining how to use the hosted DNS providers (AWS Route53/Google Cloud-DNS). We shall add this document in the upcoming release.

**Comment by shashidharatd** [comment]
Added a new item to document explaining how to use the hosted DNS providers in non-cloud environments and so reopened this issue.

**Comment by luxas** [comment]
Thanks!

**Comment by pilsy** [comment]
Any answer to https://github.com/kubernetes/kubernetes/issues/40536#issuecomment-320733514 ?

**Comment by irfanurrehman** [comment]
This issue was labelled only for sig/multicluster and is thus moved over to kubernetes/federation#101.
If this does not seem to be right, please reopen this and notify us @kubernetes/sig-multicluster-misc.
/close
